documentation/library_guide/parallel_api/pass_data_algorithms.rst

```diff
@@ -81,47 +76,65 @@ the buffer were created for the same queue. For example:
-    const int n = 1000;
-    auto policy = oneapi::dpl::execution::dpcpp_default;
-    sycl::usm_allocator<int, sycl::usm::alloc::shared> alloc(policy.queue());
-    std::vector<int, decltype(alloc)> vec(n, alloc);
-
-    std::fill(policy, vec.begin(), vec.end(), 42);
-
+    std::vector<int> vec( 1000 );
+    std::fill(oneapi::dpl::execution::dpcpp_default, vec.begin(), vec.end(), 42);
+    // each element of vec equals to 42
     return 0;
   }
 
-When using device USM, such as allocated by ``malloc_device``, manually copy data to this memory
-before calling oneDPL algorithms, and copy it back once the algorithms have finished execution.
-
-Use Host-Side std::vector
------------------------------
+When using iterators to host allocated data, a temporary SYCL buffer is created, and the data
+is copied to this buffer. After processing on a device is complete, the modified data is copied
+from the temporary buffer back to the host container. While convenient, using host allocated
+data can lead to unintended copying between host and device. We recommend working with SYCL buffers
+or USM memory to reduce data copying between the host and device. 
 
-|onedpl_long| parallel algorithms can be called with ordinary (host-side) iterators, as seen in the
-example below.
-In this case, a temporary SYCL buffer is created, and the data is copied to this buffer.
-After processing on a device is complete, the modified data is copied from the temporary buffer back
-to the host container.
-For example:
+You can also use ``std::vector`` with a USM allocator, as shown in the following example:
 
 .. code:: cpp
 
   #include <oneapi/dpl/execution>
   #include <oneapi/dpl/algorithm>
-  #include <vector>
+  #include <sycl/sycl.hpp>
   int main(){
-    std::vector<int> vec( 1000 );
-    std::fill(oneapi::dpl::execution::dpcpp_default, vec.begin(), vec.end(), 42);
-    // each element of vec equals to 42
+    const int n = 1000;
+    auto policy = oneapi::dpl::execution::dpcpp_default;
+    sycl::usm_allocator<int, sycl::usm::alloc::shared> alloc(policy.queue());
+    std::vector<int, decltype(alloc)> vec(n, alloc);
+
+    // Recommended to use USM pointers:
+    std::fill(policy, vec.data(), vec.data() + vec.size(), 42);
+
+    // Iterators for USM allocators might require extra copying
+    // std::fill(policy, vec.begin(), vec.end(), 42);
+
     return 0;
   }
 
-Working with SYCL buffers is recommended to reduce data copying between the host and device.
\ No newline at end of file
+Make sure that the execution policy and the USM-allocated memory were created for the same queue.
+
+For ``std::vector`` with a USM allocator we recommend to use ``std::vector::data()`` in
+combination with ``std::vector::size()`` as shown in the example above, rather than iterators to
+``std::vector``. That is because for some implementations of the C++ Standard Library it might not
+be possible for |onedpl_short| to detect that iterators are pointing to USM-allocated data. In that
```

Comment:
I am not sure that we should write the any internal implementation issues in details.
Other thought on that?

Reply:
This is a very high level view of the issue, and accurately describes the behavior of oneDPL, I think it is reasonable to include. 
If we omit mention of the standard library limitation, it will be difficult to fully explain what is going on here.  As we have discussed previously, it is more than just extra copies because of the limitations of host iterators (for instance as source data for `permutation_iterator`).  

Here we can accurately state that in some situations, we have no choice but to treat them as host iterators, with a reasonable explanation that this is the best we can do (rather than making it look like we just have a sloppy implementation).

Reply:
> As we have discussed previously, it is more than just extra copies because of the limitations of host iterators (for instance as source data for permutation_iterator).

I have not thought of the `permutation_iterator` problems when I coined the proposed wording. Dan, do you think it's sufficient for that case as well?

Reply:
I think without a lot of details which I don't think we want in this setting,  I don't think this will be clear to users directly.  

I think the important thing is that we recommend using USM pointers rather than iterators, and have some mention that they are treated as if they are host allocated iterators.  

With that, if users have trouble, we can point them to some additional information / known limitations to connect the dots to explain in more detail why they experience the issue they are experiencing and suggest they use USM pointers instead.